source/common/config/type_to_endpoint.cc

```diff
@@ -6,96 +6,296 @@
 
 // API_NO_BOOST_FILE
 
+#define API_VERSION envoy::config::core::v3::ApiVersion
+#define SERVICE_VERSION_INFO(v2, v3)                                                               \
+  createServiceVersionInfoMap(v2, API_VERSION::V2, {v2, v3}),                                      \
+      createServiceVersionInfoMap(v3, API_VERSION::V3, {v2, v3})
+
 namespace Envoy {
 namespace Config {
 
 namespace {
 
-// service RPC method fully qualified names.
-struct Service {
-  std::string sotw_grpc_method_;
-  std::string delta_grpc_method_;
-  std::string rest_method_;
+// A service's name, e.g. "envoy.api.v2.RouteDiscoveryService",
+// "envoy.service.route.v3.RouteDiscoveryService".
+using ServiceName = std::string;
+
+struct ServiceVersionInfo {
+  // The version of a service. Possible values: envoy::config::core::v3::ApiVersion::V2,
+  // envoy::config::core::v3::ApiVersion::V3.
+  envoy::config::core::v3::ApiVersion transport_api_version_;
+
+  // This hold a name for each version, for example for "envoy.api.v2.RouteDiscoveryService".
+  //
+  // {
+  //    "V2": "envoy.api.v2.RouteDiscoveryService",
+  //    "V3": "envoy.service.route.v3.RouteDiscoveryService"
+  // }
+  std::unordered_map<envoy::config::core::v3::ApiVersion, ServiceName> names_;
+};
+
+// A ServiceVersionInfoMap holds a service's version and possible names for each available version.
+// For examples:
+//
+// Given "envoy.api.v2.RouteDiscoveryService" as the service name:
+// {
+//   "envoy.api.v2.RouteDiscoveryService": {
+//     "transport_api_version_": "V2",
+//     "names_": {
+//       "V2": "envoy.api.v2.RouteDiscoveryService",
+//       "V3": "envoy.service.route.v3.RouteDiscoveryService"
+//     }
+//   }
+// }
+//
+// And for "envoy.service.route.v3.RouteDiscoveryService":
+// {
+//   "envoy.service.route.v3.RouteDiscoveryService": {
+//     "transport_api_version_": "V3",
+//     "names_": {
+//       "V2": "envoy.api.v2.RouteDiscoveryService",
+//       "V3": "envoy.service.route.v3.RouteDiscoveryService"
+//     }
+//   }
+// }
+using ServiceVersionInfoMap = std::unordered_map<ServiceName, ServiceVersionInfo>;
+
+// This creates a ServiceToServiceVersionInfoMap, with service name (e.g.
```

Comment:
This returns a `ServiceVersionInfoMap`, not `ServiceToServiceVersionInfoMap`

Reply:
Thanks for catching this! Updated.